### PR TITLE
Log to specific instead of working directory

### PR DESCRIPTION
**Pull request by sanmai-NL**
- Add a configuration setting `logs_dir`/`PDM_LOGS_DIR`.
- Improve handling of temporary file to guard against TOCTOU defects.
- Reduce couping between `termui` and `core` to avoid cyclic dependency.

## Pull Request Checklist

- [ ] A news fragment is added in `news/` describing what is new.
- [ ] Test cases added for changed code.

### CODE REVIEW

**Comment by sanmai-NL** [comment]
#2398
@frostming @pawamoy after your review, I'll add the news fragment. Not sure whether and how to test this.

File: src/pdm/project/config.py [hunk 1]
```diff
@@ -338,7 +349,7 @@ def __setitem__(self, key: str, value: Any) -> None:
 
         value = config.coerce(value)
         if key in self.env_map:
-            ui.warn(f"the config is shadowed by env var '{config.env_var}', the value set won't take effect.")
+            raise PDMWarning(f"The config is shadowed by env var '{config.env_var}', the value set won't take effect.")
```
Comment (frostming) [inline src/pdm/project/config.py:352]: Why this change? this will cause a hard fail which is not expected.
